zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,520 @@
+    q_word_start: Selector, // enable when word starts
+    input: Column<Advice>,  // input bitstream byte/row
+    // output parts:
+    // encoded value
+    data: Column<Advice>,
+    data_rlc: Column<Advice>,
+    // encoded bit length
+    num_bits: Column<Advice>,
+    // randomness multiplier selector
+    r_multi: Column<Advice>,
+    // the factor with which the part needs to be multiplied
+    shift_factor: Column<Advice>,
+    // the factor with which the part needs to be multiplied when
+    // we decode it
+    decode_factor: Column<Advice>,
+    // fixed huffman table
+    huffman_table: [TableColumn; 3], // [enable, value, code, code_length]
+
+    _marker: PhantomData<F>,
+}
+
+/// Circuit configuration arguments
+pub struct CompressCircuitConfigArgs<F> {
+    // randomness used to do RLC
+    pub randomness: Expression<F>,
+}
+
+impl<F: Field> SubCircuitConfig<F> for CompressCircuitConfig<F> {
+    type ConfigArgs = CompressCircuitConfigArgs<F>;
+
+    fn new(
+        meta: &mut ConstraintSystem<F>,
+        CompressCircuitConfigArgs { randomness }: Self::ConfigArgs,
+    ) -> Self {
+        let q_end = meta.selector();
+        let q_not_end = meta.selector();
+        let q_word_start = meta.complex_selector();
+        let r_multi = meta.advice_column();
+        let input = meta.advice_column();
+        let data = meta.advice_column();
+        let data_rlc = meta.advice_column();
+        let num_bits = meta.advice_column();
+        let shift_factor = meta.advice_column();
+        let decode_factor = meta.advice_column();
+        let huffman_table = array_init::array_init(|_| meta.lookup_table_column());
+
+        meta.enable_equality(data_rlc);
+
+        let mut parts = vec![];
+        meta.create_gate("split parts", |meta| {
+            // next 4 rows as 4 parts
+            for i in 0..MAX_PARTS as i32 {
+                let data = meta.query_advice(data, Rotation(i));
+                let num_bits = meta.query_advice(num_bits, Rotation(i));
+                let r_multi = meta.query_advice(r_multi, Rotation(i));
+                let shift_factor = meta.query_advice(shift_factor, Rotation(i));
+                let decode_factor = meta.query_advice(decode_factor, Rotation(i));
+
+                parts.push(Part {
+                    data,
+                    num_bits,
+                    r_multi,
+                    shift_factor,
+                    decode_factor,
+                });
+            }
+            vec![0.expr()]
+        });
+
+        meta.lookup("huffman table", |meta| {
+            // lookups:
+            //  1. input = huffman_value
+            //  2. data0 * decode_factor0 + data1 * decode_factor1 + data2 * decode_factor2
+            //      + data3 * decode_factor3 = huffman_code
+            //  3. num_bits0 + num_bits1 + num_bits2 + num_bits3 = huffman_code_length
+            let input = meta.query_advice(input, Rotation::cur());
+            let code = decode::data_expr(&parts);
```

Comment:
As you're doing the part/row route, you'll have to be careful to correctly construct your table to return the correct data for each row between 0..MAX_PARTS. Perhaps you'll need to add some row_id to the lookup table or something like that. For example, because you're planning to decode the parts like `data0 * decode_factor0 + data1 * decode_factor1  + data3 * decode_factor3`, `decode_factor` may be row dependent, though not sure. Also the decoding could perhaps be done differently so this isn't necessary.

Reply:
Yes, it is a little difficult here. I think it works expected by testcase
```rust
 				match code_len.cmp(&remaining) {
                    std::cmp::Ordering::Less => {
                        num_bits = code_len;
                        shift_factor = 2u8.pow((remaining - num_bits) as u32);
                        decode_factor = 1;
                        remaining -= code_len;
                        next_byte_begin = false;
                        code_len = 0;
                    }
                    std::cmp::Ordering::Equal => {
                        num_bits = code_len;
                        shift_factor = 1;
                        decode_factor = 1;
                        remaining = 8; // next byte begin
                        next_byte_begin = true;
                        code_len = 0;
                    }
                    std::cmp::Ordering::Greater => {
                        num_bits = remaining;
                        shift_factor = 1;
                        decode_factor = 2u32.pow((code_len - num_bits) as u32);
                        code_len -= remaining;
                        remaining = 8; // next byte begin
                        next_byte_begin = true;
                    }
                }
```